File: ios/TransportHQ/Models/RideRequest.swift
import Foundation

struct RideRequest: Codable, Identifiable {
    let id: String
    let passenger: TripPassenger?
    let requester: TripPassenger?
    let pickupLocation: FlexLocation?
    let dropoffLocation: FlexLocation?
    let pickupTime: String?
    let passengerCount: Int?
    let priority: RidePriority?
    let status: RideRequestStatus?
    let department: String?
    let notes: String?
    let driver: TripDriver?
    let vehicle: TripVehicle?
    let assignedDriver: TripDriver?
    let assignedVehicle: TripVehicle?
    let trip: String?
    let eta: String?
    let isPACall: Bool?
    let callType: String?
    let isSharedRide: Bool?
    let createdAt: String?
    let updatedAt: String?

    enum CodingKeys: String, CodingKey {
        case id = "_id"
        case passenger, requester, pickupLocation, dropoffLocation, pickupTime
        case passengerCount, priority, status, department, notes
        case driver, vehicle, assignedDriver, assignedVehicle
        case trip, eta, isPACall, callType, isSharedRide, createdAt, updatedAt
    }

    var displayPassenger: TripPassenger? { passenger ?? requester }
    var displayDriver: TripDriver? { driver ?? assignedDriver }
    var displayVehicle: TripVehicle? { vehicle ?? assignedVehicle }
}

// Handles both String and Object location formats from API
struct FlexLocation: Codable {
    let address: String
    let coordinates: GeoPoint?

    init(from decoder: Decoder) throws {
        if let container = try? decoder.singleValueContainer(), let str = try? container.decode(String.self) {
            self.address = str
            self.coordinates = nil
        } else {
            let container = try decoder.container(keyedBy: CodingKeys.self)
            self.address = (try? container.decode(String.self, forKey: .address)) ?? ""
            self.coordinates = try? container.decode(GeoPoint.self, forKey: .coordinates)
        }
    }

    enum CodingKeys: String, CodingKey {
        case address, coordinates
    }
}

enum RidePriority: String, Codable {
    case low, normal, high, urgent

    var displayName: String { rawValue.capitalized }
}

enum RideRequestStatus: String, Codable {
    case pending
    case approved
    case assigned
    case inProgress = "in-progress"
    case completed
    case cancelled
    case rejected

    var displayName: String {
        switch self {
        case .pending: return "Pending"
        case .approved: return "Approved"
        case .assigned: return "Assigned"
        case .inProgress: return "In Progress"
        case .completed: return "Completed"
        case .cancelled: return "Cancelled"
        case .rejected: return "Rejected"
        }
    }
}

struct CreateRideRequest: Codable {
    let pickupLocation: String
    let dropoffLocation: String
    let pickupTime: String
    let passengerCount: Int
    let priority: String
    let department: String
    let notes: String
}

File: android/app/src/main/java/com/transporthq/app/data/models/RideRequest.kt
package com.transporthq.app.data.models

import com.google.gson.annotations.SerializedName

data class RideRequest(
    @SerializedName("_id") val id: String = "",
    val passenger: PassengerRef? = null,
    val requester: PassengerRef? = null,
    val pickupLocation: String? = null,
    val dropoffLocation: String? = null,
    val pickupTime: String? = null,
    val passengerCount: Int = 1,
    val priority: String = "normal",
    val status: String = "pending",
    val department: String = "",
    val notes: String? = null,
    val driver: DriverRef? = null,
    val vehicle: VehicleRef? = null,
    val assignedDriver: DriverRef? = null,
    val assignedVehicle: VehicleRef? = null,
    val isPACall: Boolean = false,
    val callType: String? = null,
    val isSharedRide: Boolean = false,
    val eta: String? = null,
    val trip: String? = null,
    val createdAt: String? = null,
    val updatedAt: String? = null
) {
    val displayDriver: DriverRef? get() = driver ?: assignedDriver
    val displayVehicle: VehicleRef? get() = vehicle ?: assignedVehicle
    val displayPassenger: PassengerRef? get() = passenger ?: requester
}

data class CreateRideRequest(
    val pickupLocation: String,
    val dropoffLocation: String,
    val pickupTime: String,
    val passengerCount: Int,
    val priority: String = "normal",
    val department: String = "",
    val notes: String = ""
)

data class RideStatusUpdate(
    val status: String
)

data class AssignRideRequest(
    val driverId: String,
    val vehicleId: String,
    val eta: Int = 15
)

data class RideRequestsResponse(
    val rideRequests: List<RideRequest> = emptyList()
)

data class RideRequestResponse(
    val rideRequest: RideRequest
)

data class RejectRideRequest(
    val reason: String
)
